{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from collections import OrderedDict\n",
    "\n",
    "class BAMHeader (object):\n",
    "    \n",
    "    def __init__ (self):          \n",
    "        \n",
    "        # Init the header dict with a minimal header line\n",
    "        self.header = {'HD':{'VN': '1.5', 'SO':\"unknown\"}, 'SQ':[]}\n",
    "        \n",
    "        # Init enpty dict fo conversion between rname and refid \n",
    "        self.name_to_id = OrderedDict()\n",
    "        self.id_to_name = OrderedDict()\n",
    "        self.current_id = 0\n",
    "    \n",
    "    def add_header_line (self, line):\n",
    "            \n",
    "        # Try except bloc to skip to the next line in case a non standard line is found\n",
    "        try:\n",
    "            split_line = line.strip().split(\"\\t\")\n",
    "            tag = split_line[0].lstrip(\"@\")\n",
    "\n",
    "            # In case a HD tag is found will replace the minimal header\n",
    "            if tag == \"HD\":\n",
    "                for field in split_line[1:]:\n",
    "                    key, value = field.split(\":\")\n",
    "                    self.header[\"HD\"][key] = value\n",
    "\n",
    "            # In case a SQ tag is found = build a dict to have correspondance between id and refname\n",
    "            elif tag == \"SQ\":\n",
    "                d={}\n",
    "                for field in split_line[1:]:\n",
    "                    key, value = field.split(\":\")\n",
    "                    d[key] = int(value) if value.isdigit() else value\n",
    "\n",
    "                # append to the list of SQ lines\n",
    "                self.header[tag].append(d)\n",
    "\n",
    "                # append to the dict of correspondance\n",
    "                self.name_to_id [d['SN']] = self.current_id\n",
    "                self.id_to_name [self.current_id] = d['SN']\n",
    "                self.current_id += 1\n",
    "\n",
    "            # For all the other tags, create an entry with a list of values found\n",
    "            else:\n",
    "                d ={}\n",
    "                for field in split_line[1:]:\n",
    "                    key, value = field.split(\":\")\n",
    "                    d[key] = int(value) if value.isdigit() else value\n",
    "                if tag not in self.header:\n",
    "                    self.header[tag]=[d]\n",
    "                else:\n",
    "                    self.header[tag].append(d)\n",
    "\n",
    "        except Exception as E:\n",
    "            print E\n",
    "            print \"Invalid sequence line in bam: {}\".format(line)\n",
    "    \n",
    "    def rname_to_refid (self, name):\n",
    "        \"\"\"return the id corresponding to the reference sequence name (pysam compatible)\"\"\"\n",
    "        if name == \"*\":\n",
    "            return -1\n",
    "        else:\n",
    "            return self.name_to_id[name]\n",
    "\n",
    "    def refid_to_refname (self, id):\n",
    "        \"\"\"return the name corresponding to the reference sequence id (pysam compatible)\"\"\"\n",
    "        if id == -1:\n",
    "            return \"*\"\n",
    "        else:\n",
    "            return self.id_to_name[id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pysam\n",
    "from collections import OrderedDict\n",
    "from re import compile as recompile\n",
    "\n",
    "class BAMSequenceParser (object):\n",
    "       \n",
    "    # Init method\n",
    "    def __init__ (self, header, skip_secondary=True):\n",
    "        \n",
    "        # Store defined variables\n",
    "        self.header = header\n",
    "        self.skip_secondary = skip_secondary\n",
    "        \n",
    "        # counters\n",
    "        self.count = {\"all\":0, \"invalid\":0, \"primary\":0, \"secondary\":0}\n",
    "\n",
    "        # For cigar to tuple code conversion\n",
    "        self.cigar_to_code = {'M':0, 'I':1, 'D':2, 'N':3, 'S':4, 'H':5, 'P':6, '=':7, 'X':8}\n",
    "        self.cigar_regex = recompile(\"(\\d+)([MIDNSHP=X])\")\n",
    "        \n",
    "    def __str__ (self):\n",
    "        return \"\".join([\"{}\\t{}\\n\".format(i,j) for i, j in self.count.items()])\n",
    "        \n",
    "    def parse_line (self, line):\n",
    "    \n",
    "        self.count[\"all\"] += 1\n",
    "        # Try bloc to skip to the next line in case a non standard line is found\n",
    "        try:\n",
    "            # Split the line and verify the number of fields \n",
    "            split_line = line.strip().split(\"\\t\")\n",
    "            assert len (split_line) >=  11, \"Invalid number of field in bam aligned sequence\"\n",
    "      \n",
    "            # Init an AlignedSegment\n",
    "            read = Read(\n",
    "                qname = split_line[0],\n",
    "                flag = int(split_line[1]),\n",
    "                rname = self.header.rname_to_refid(split_line[2]),\n",
    "                pos = int(split_line[3])-1,\n",
    "                mapq = int(split_line[4]),\n",
    "                cigar = self.cigarstring_to_tuple(split_line[5]),\n",
    "                rnext = self.header.rname_to_refid(split_line[6]),\n",
    "                pnext = int(split_line[7])-1,\n",
    "                seq = split_line[9],\n",
    "                qual = pysam.fromQualityString(split_line[10]),\n",
    "                tags = self.parse_tags(split_line [11:]) if len (split_line) >= 12 else tuple())\n",
    "            \n",
    "            # skip the read if secondary and required\n",
    "            if read.is_secondary:\n",
    "                self.count[\"secondary\"] += 1\n",
    "                if self.skip_secondary:\n",
    "                    return None\n",
    "            \n",
    "            # finally return the read\n",
    "            self.count[\"primary\"] += 1\n",
    "            return read\n",
    "        \n",
    "        except Exception as E:\n",
    "            print E\n",
    "            print \"Invalid sequence line in bam: {}\".format(line)\n",
    "            self.count[\"invalid\"] += 1\n",
    "            return None\n",
    "                \n",
    "    def cigarstring_to_tuple (self, cigarstring):\n",
    "        \"\"\"Transform the cigar string in a pysam compatible tuple\"\"\"\n",
    "        if cigarstring == \"*\":\n",
    "            return None\n",
    "        else:\n",
    "            return tuple([(self.cigar_to_code[y], int(x)) for x, y in self.cigar_regex.findall(cigarstring)])\n",
    "    \n",
    "    \n",
    "    def parse_tags (self, tags_split):\n",
    "        \"\"\"Parse the optional tag fields\"\"\"\n",
    "        tags = []\n",
    "        \n",
    "        for tag in tags_split:\n",
    "            tag_split = tag.split(\":\")\n",
    "            key = tag_split[0]\n",
    "            value = int(tag_split[2]) if tag_split[1] == 'i' else tag_split[2]\n",
    "            tags.append((key, value,))\n",
    "            \n",
    "        return tuple(tags)\n",
    "    \n",
    "class Read (object):\n",
    "    \"\"\"Wrapping class over a pysam AlignedSegment to transform its initialisation\"\"\"\n",
    "    \n",
    "    def __init__ (self, qname, flag, rname, pos, mapq, cigar, rnext, pnext, seq, qual, tags):\n",
    "        \"\"\"Create a pysam.AlignedSegment object and define fields\"\"\"\n",
    "        \n",
    "        self.read = pysam.AlignedSegment()\n",
    "        self.read.query_name = qname\n",
    "        self.read.flag = flag\n",
    "        self.read.reference_id = rname\n",
    "        self.read.reference_start = pos\n",
    "        self.read.mapping_quality = mapq\n",
    "        self.read.cigar = cigar\n",
    "        self.read.next_reference_id = rnext\n",
    "        self.read.next_reference_start = pnext\n",
    "        self.read.template_length = self.read.infer_query_length() if self.read.cigar else 0\n",
    "        self.read.query_sequence = seq\n",
    "        self.read.query_qualities = qual\n",
    "        self.read.tags = tags\n",
    "        \n",
    "        self.is_secondary = self.read.is_secondary\n",
    "        \n",
    "    def is_properly_mapped(self, min_mapq, min_match_size):\n",
    "        \"\"\"Is properly mapped if mapped on e sequence with a mapq score higher the min_mapq\n",
    "        and a template lenght higher than min_match_size\"\"\"\n",
    "        return (self.read.tid != -1 and self.read.mapq >= min_mapq and self.read.template_length >= min_match_size)\n",
    "    \n",
    "    def to_bam (self):\n",
    "        return self.read\n",
    "    \n",
    "    def to_fastq (self):\n",
    "        return \"{}\\n{}\\n+\\n{}\\n\".format(self.read.qname, self.read.seq, self.read.qual)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "need more than 1 value to unpack\n",
      "Invalid sequence line in bam: @CO\tInvalid line\n",
      "\n",
      "too many values to unpack\n",
      "Invalid sequence line in bam: @SQ\tSN:invalid_seq\tLN:227:57\n",
      "\n",
      "all\t11\n",
      "primary\t10\n",
      "invalid\t0\n",
      "secondary\t1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pysam\n",
    "from gzip import open as gopen \n",
    "\n",
    "min_mapq = 10\n",
    "min_match_size = 25\n",
    "\n",
    "with open (\"./test/test.sam\", \"r\") as sam:\n",
    "    \n",
    "    # Parse bam header\n",
    "    h = BAMHeader ()\n",
    "    for line in sam:\n",
    "        \n",
    "        # Add to BAMheader object if the line is a header line \n",
    "        if line.startswith(\"@\"):\n",
    "            h.add_header_line(line)\n",
    "        \n",
    "        # Exit when at end of header\n",
    "        else:\n",
    "            break\n",
    "    \n",
    "    # Create an output bam file for mapped reads and an ouput fastq file for unmapped reads \n",
    "    with \\\n",
    "        pysam.AlignmentFile(\"out.sam\", \"wh\", header=h.header) as bam,\\\n",
    "        open (\"out.fastq\", \"w\") as fastq:\n",
    "        \n",
    "        # Initialize a bam sequence parser\n",
    "        bam_parser = BAMSequenceParser (header = h, skip_secondary = True)\n",
    "        \n",
    "        # Process the first sequence found when parsing header \n",
    "        read = bam_parser.parse_line(line)\n",
    "        if read:\n",
    "            if read.is_properly_mapped(min_mapq, min_match_size):\n",
    "                bam.write(read.to_bam())\n",
    "            else:\n",
    "                fastq.write(read.to_fastq())\n",
    "        \n",
    "        # Process the remaining sequences        \n",
    "        for line in sam:\n",
    "            read = bam_parser.parse_line(line)\n",
    "            if read:\n",
    "                if read.is_properly_mapped(min_mapq, min_match_size):\n",
    "                    bam.write(read.to_bam())\n",
    "                else:\n",
    "                    fastq.write(read.to_fastq())\n",
    "                    \n",
    "print (bam_parser)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
